### PR TITLE
Geoopt

### PR DESCRIPTION
**Pull request by baumeier**
Moved everything that dealt with geometry optimization of excited states (numerical gradients, optmization algorithm) out of exciton.h/.cc to separate objects.

### CODE REVIEW

File: src/libxtp/forces.cc [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+            // checking if there is only one segment
+            _nsegments = _segments.size();
+            if (_nsegments > 1) throw runtime_error(string("\n Force calculation for more than 1 conjugated segment not supported. Stopping!"));
+
+            // pre-check forces method
+            std::vector<string> choices = {"forward", "central"};
+            _force_method = options->ifExistsAndinListReturnElseThrowRuntimeError<string>(".method", choices);
+
+            if ((_force_method == "forward") || (_force_method == "central")) {
```
Comment (JensWehner) [inline src/libxtp/forces.cc]: does not need if clause

File: src/libxtp/forces.cc [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+                (*ait)->setQMPos(_current_pos); // restore original coordinate into segment
+            }
+
+            return;
+        }
+
+
+
+void Forces::Orbitals2Segment(ctp::Segment* _segment, Orbitals* _orbitals) {
```
Comment (JensWehner) [inline src/libxtp/forces.cc]: Should this maybe not be part of orbitals?

File: src/libxtp/geometry_optimization.cc [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+                votca::tools::vec pos = (*ait)->getQMPos();
+                string name = (*ait)->getElement();
+
+                fprintf(out, "%2s %4.7f %4.7f %4.7f atom_vector %4.7f %4.7f %4.7f \n",
+                        name.c_str(),
+                        pos.getX()*10,
+                        pos.getY()*10,
+                        pos.getZ()*10,
+                        _force(_i_atom, 0)*51.4220822067,
```
Comment (JensWehner) [inline src/libxtp/geometry_optimization.cc]: Explain magic constant

File: src/libxtp/gwbse/gwbseengine.cc [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            // write logger to log file
+            ofstream ofs;
+            //string gwbse_logfile = "gwbse.log";
+            ofs.open(_logger_file.c_str(), ofstream::out);
+            if (!ofs.is_open()) {
+                throw runtime_error("Bad file handle: " + _logger_file);
+            }
+            ofs << (*pLog) << endl;
+            ofs.close();
+
```
Comment (JensWehner) [inline src/libxtp/gwbse/gwbseengine.cc]: return Statement and more explicit name for DumpLog, what does it do exactly

File: src/libxtp/tools/exciton.cc [hunk 5]
```diff
@@ -16,1121 +16,154 @@
...
-    
-    
-    ub::vector<double> _new_pos = _current_pos + _delta_pos;
-    CTP_LOG(ctp::logDEBUG,_log) << "BFGS-TRM: step " << sqrt(_norm_delta_pos) << " vs TR " << sqrt(_trust_radius_squared) << flush  ;
 
-    // get the energy estimate for a local quadratic surface
-    ub::vector<double> _temp_ene = ub::prod( _hessian, _delta_pos );
-    _delta_energy_estimate = ub::inner_prod( _current_gradient, _delta_pos ) + 0.5 * ub::inner_prod(_delta_pos,_temp_ene );
+            CTP_LOG(ctp::logDEBUG, _log) << "Saving data to " << _archive_file << flush;
+            std::ofstream ofs((_archive_file).c_str());
```
Comment (JensWehner) [inline src/libxtp/tools/exciton.cc]: why not use the orbital method to save to file?

File: src/libxtp/tools/exciton.cc [hunk 6]
```diff
@@ -16,1121 +16,154 @@
...
-             //int _idx = 3*_i_atom + _i_cart;
-            //_current_xyz(_i_atom,_i_cart) -= _total_shift(_i_cart)/_natoms;
-            // _trial_xyz(_i_atom,_i_cart) -= _total_shift(_i_cart)/_natoms;
-        }
-    }
-    */
-    return;
-}
 
+        void Exciton::ReadXYZ(ctp::Segment* _segment, string filename) {
```
Comment (JensWehner) [inline src/libxtp/tools/exciton.cc]: This should be part of ctp::Segment but okay

File: src/libxtp/tools/exciton.h [hunk 7]
```diff
@@ -24,121 +24,72 @@
...
+            bool _do_optimize;
+
+
+            void ReadXYZ(ctp::Segment* _segment, string filename);
+            
+            void PrepareGuess(Orbitals *_orbitalsA,
+                    Orbitals *_orbitalsB,
+                    Orbitals *_orbitalsAB);
+
+            void OrthonormalizeGuess(ctp::Segment* _segment, Orbitals* _orbitals);
```
Comment (JensWehner) [inline src/libxtp/tools/exciton.h]: does this function still exist? Maybe remove?

File: src/libxtp/forces.cc (hunk 2 above)
Comment (baumeier) [inline src/libxtp/forces.cc]: went to qminterface

File: src/libxtp/geometry_optimization.cc (hunk 3 above)
Comment (baumeier) [inline src/libxtp/geometry_optimization.cc]: The whole thing should have been gone. Was output for a special program.

File: src/libxtp/gwbse/gwbseengine.cc (hunk 4 above)
Comment (baumeier) [inline src/libxtp/gwbse/gwbseengine.cc]: Done.

**Pull request merged**